x/upgrade/internal/keeper/keeper.go

```diff
@@ -0,0 +1,119 @@
+// with this name is applied. In order for an upgrade with the given name to proceed, a handler for this upgrade
+// must be set even if it is a no-op function.
+func (k *Keeper) SetUpgradeHandler(name string, upgradeHandler types.UpgradeHandler) {
+	k.upgradeHandlers[name] = upgradeHandler
+}
+
+// ScheduleUpgrade schedules an upgrade based on the specified plan.
+// It there is another Plan already scheduled, it will overwrite it
+// (implicitly cancelling the current plan)
+func (k *Keeper) ScheduleUpgrade(ctx sdk.Context, plan types.Plan) sdk.Error {
+	err := plan.ValidateBasic()
+	if err != nil {
+		return err
+	}
+	if !plan.Time.IsZero() {
+		if !plan.Time.After(ctx.BlockHeader().Time) {
+			return sdk.ErrUnknownRequest("upgrade cannot be scheduled in the past")
+		}
+	} else if plan.Height <= ctx.BlockHeight() {
+		return sdk.ErrUnknownRequest("upgrade cannot be scheduled in the past")
+	}
+
+	store := ctx.KVStore(k.storeKey)
+	if store.Has(types.DoneHeightKey(plan.Name)) {
+		return sdk.ErrUnknownRequest(fmt.Sprintf("upgrade with name %s has already been completed", plan.Name))
+	}
+
+	bz := k.cdc.MustMarshalBinaryBare(plan)
+	store.Set(types.PlanKey(), bz)
+	return nil
+}
+
+func (k *Keeper) getDoneHeight(ctx sdk.Context, name string) int64 {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.DoneHeightKey(name))
+	if len(bz) == 0 {
+		return 0
+	}
+	return int64(binary.BigEndian.Uint64(bz))
+}
+
+// ClearUpgradePlan clears any schedule upgrade
+func (k *Keeper) ClearUpgradePlan(ctx sdk.Context) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.PlanKey())
+}
+
+// Logger returns a module-specific logger.
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+// GetUpgradePlan returns the currently scheduled Plan if any, setting havePlan to true if there is a scheduled
+// upgrade or false if there is none
+func (k *Keeper) GetUpgradePlan(ctx sdk.Context) (plan types.Plan, havePlan bool) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.PlanKey())
+	if bz == nil {
+		return plan, false
+	}
+	k.cdc.MustUnmarshalBinaryBare(bz, &plan)
+	return plan, true
+}
+
+// setDone marks this upgrade name as being done so the name can't be reused accidentally
+func (k *Keeper) setDone(ctx sdk.Context, name string) {
+	store := ctx.KVStore(k.storeKey)
+	bz := make([]byte, 8)
+	binary.BigEndian.PutUint64(bz, uint64(ctx.BlockHeight()))
+	store.Set(types.DoneHeightKey(name), bz)
+}
+
+// HasHandler returns true iff there is a handler registered for this name
+func (k *Keeper) HasHandler(name string) bool {
+	_, ok := k.upgradeHandlers[name]
+	return ok
+}
+
+// ApplyUpgrade will execute the handler associated with the Plan and mark the plan as done.
+func (k *Keeper) ApplyUpgrade(ctx sdk.Context, plan types.Plan) {
+	handler := k.upgradeHandlers[plan.Name]
```

Comment:
See above. when and where is this set?

Reply:
It is stored in memory. Depends on the `binary` not the db state.

This is the key point we use to differentiate two different versions of the binary - if they have the handler for XYZ installed, which must be done in App Initialization, before processing the first block.